### ISSUE TITLE
Feature Request: Kubernetes Operator

### ISSUE THREAD

**Issue by toonsevrin** [tailscale/tailscale#502]
I may implement this myself, so consider this a tracker.

Essentially I want to be able to locally access kubernetes services (in an already secured cluster) within the security of Tailscale. We have a ton of corporate services that are currently acessed with "kubectl port-forward", using tailscale would streamline this a lot (dns and no command line).

My preference goes to having a tailscale identity (IP) per services you wish to expose to the network. This can be achieved by having an operator that watches LoadBalancer Services. If they are labelled with "tailscale: true", the operator creates an identity and a proxy deployment that terminates the tailscale traffic for that identity, routing the traffic to the actual service.

What would be the best way to handle identity in this setup?

**Comment by danderson** [comment]
We're not going to work on this just yet, blocked on #504. The questions you're asking are exactly the ones we need to answer before building a Kubernetes operator, and we haven't thought about it at all yet. If you're interested in building this yourself, you're welcome to experiment with the tradeoffs and see what works best in the k8s ecosystem.

**Comment by toonsevrin** [comment]
Once I do need this, I'll implement it myself :)

Can you perhaps provide a bit more detail in #504 at the issues that you are facing with containerization, the potential solutions and why they are not immediately solveable?

Thank you very much!

**Comment by zfogg** [comment]
i really want to see this. please do it

**Comment by bithavoc** [comment]
This feature would be absolutely amazing, imagine using Tailscale for cross-cluster resource access 🥇 

There is money to be made here Tailscale, please take my money.

**Comment by niels-s** [comment]
We recently deployed Tailscale on our k8s cluster, we had to tweak our CoreDNS config a bit, but right now, when DNS is enabled, you can do something like `curl http://<svc name.<namespace>.svc.osig.lan`. 
It's pretty minimal and doesn't come with cross-cluster connection or per-service settings like requested above. My colleague @alex-richman-onesignal  is writing a blog post on it, pinging him to share the story once published.

**Comment by enigmatix** [comment]
I'd be very interested to learn exactly how you did it and if there were any issues.  I'm currently working through an implementation myself @niels-s @alex-richman-onesignal

**Comment by bithavoc** [comment]
@niels-s how did you get your pods to talk to the MagicDNS Server in first place?

**Comment by bithavoc** [comment]
So I got MagicDNS working inside Kubernetes(K3S, Ubuntu 20.04), all I did was a simple edit to the ConfigMap for CoreDNS to prepend a new server just for `beta.tailscale.net` forwarding everything to `100.100.100.100`(the nameserver IP of all users in the Beta) like this:


```
beta.tailscale.net:53 {
        loop
        cache 10
        forward . 100.100.100.100
    }
// ... the rest of the CoreDNS config file...
```

So my config map now looks as follows:

```
data:
  Corefile: |
    beta.tailscale.net:53 {
        loop
        cache 10
        forward . 100.100.100.100
    }
    .:53 {
        errors
        health
        ready
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        hosts /etc/coredns/NodeHosts {
          ttl 60
          reload 15s
          fallthrough
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```

Once I saved the config file, I reloaded the CoreDNS pods so they pick the new config map:

```
kubectl rollout restart -n kube-system deployment/coredns
```

And that's all I did, I was able to use MagicDNS from any pod in the cluster.

Now that with this setup, you can use MagicDNS **only** with FQDNS such as `<hostname>.<tailscale-domain>.beta.tailscale.net`.

If you want to use just the `hostname` and have it resolve to tailscale MagicDNS, one approach would be to just use `forward . 100.100.100.1000` in your `.:53` server but I wouldn't recommend that since it may mess with the resolution of other services in the networking your cluster might depend on.

**Comment by DentonGentry** [comment]
Related: https://github.com/tailscale/tailscale/issues/2782

**Comment by timothyclin** [comment]
I don't think that is addressed in the referenced. The main ask from OP is to have tailscale act as either ingress or loadbalancer service automatically. The quoted feature is not automatic and relies on hand configuring each and every service/pods spun-up. 

> I think this is addressed by https://github.com/maisem/tailscale-on-k8s (which may move into the tailscale organization at some point for visibility, Maisem is a software engineer at Tailscale).

**Comment by maisem** [comment]
+1 to what @timothyclin said. The ask here is to have something that automatically spins up a Tailscale proxy instead of having a user manually do it. The proxy itself would be something like https://github.com/tailscale/tailscale/tree/main/docs/k8s#sample-proxy.

**Comment by maisem** [comment]
The thing is that there are multiple use cases for running Tailscale on Kubernetes, as evident from our [k8s docs](https://github.com/tailscale/tailscale/tree/main/docs/k8s), which don't list all of them either.

Writing an operator would most likely involve adding multiple controllers not just one. But mostly, the use cases can be broken down into
1. Providing inbound connectivity to a group of pods (or service)
    This is basically sharing something running inside a cluster to Tailscale devices over a Tailscale IP.
    
    This could be accomplished by adding a controller which watches services annotated with a special key, and creates a StatefulSet with `replicas: 1`. The actual pod itself would be a variant of [sample-proxy](https://github.com/tailscale/tailscale/tree/main/docs/k8s#sample-proxy).
    Increasing the replica count beyond `1` would most likely require #465, or some workaround using subnet routers and externally assigned VIPs.
2. Providing inbound connectivity to a single pod e.g. for StatefulSet pods to communicate with each other
3. Providing outbound connectivity from pods 
    a. to Tailscale devices
    b. other pods in the same cluster
    c. cross-cluster connectivity

     The obvious solution for 2&3 would be to automatically inject a Tailscale sidecar for each pod annotated with something like `ts.net/enable=true` with a help of a [Mutating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook). This has a few caveats though which makes this non-ideal.
     - It could potentially leak a lot of Tailscale nodes if not used carefully. This is something that [Ephemeral nodes](https://tailscale.com/kb/1111/ephemeral-nodes/) should help with.
     - It'll end up potentially running many Tailscale nodes on the same machine, which would be expensive and wasteful. OTOH it would provide the ability to write pod-level [ACL](https://tailscale.com/kb/1018/acls/) policies with the help of [ACL tags](https://tailscale.com/kb/1068/acl-tags/).

    Another approach would be to run Tailscale on the node instead. Each `tailscaled` would then advertise each nodes [podCIDRs](https://pkg.go.dev/k8s.io/api/core/v1#NodeSpec.PodCIDRs) as a [subnet route](https://tailscale.com/kb/1019/subnets/). This would still allow outbound access and cross-cluster networking, however the configuration would still require additional [NetworkPolicy](https://kubernetes.io/docs/concepts/services-networking/network-policies/) management.
    As long as Tailscale on the node is configured to `--accept-routes`, Pod<->Pod communication should still occur over Tailscale. However, the traffic would not be encrypted until it's about to leave the node, which means that it will be in plaintext on the node itself. (We should make sure that this configuration works on GKE/EKS/etc with their respective CNIs.)
    
    This is probably achievable using a simple DaemonSet with the a [subnet-router pod](https://github.com/tailscale/tailscale/tree/main/docs/k8s#subnet-router) in hostNetworking mode.

    For self-hosted k8s clusters, we could also provide a CNI which could assign node-local IPs like the old days of [kubenet](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/#kubenet) and run Tailscale on the node.

**Comment by halkeye** [comment]
As requested on twitter - https://twitter.com/maisem_ali/status/1498087409187057664 - my use cases:

The majority of my use cases are to work around the limitations of helm.  I want to be able to deploy jenkins to my cloud based k8s server, but still have access to the tiny infra I have in my apartment. [The helm chart supports adding side cars](https://github.com/jenkinsci/helm-charts/blob/9752f8a19f5bb39183821713d7b47e4a17a61c7d/charts/jenkins/values.yaml#L377), but having to manage roles, and secrets, and everything else makes this complicated again.

my other immediate use case would be installing the overseearr dashboard in the cloud, talking to services running on my nas. The helm chart again semi supports sidecars, but not additional roles.

The third, less likely, but scenario i have is setting up portainer/pterodactyl or in the cloud, and having agents running on my nas or cloud vms that connect back to the server. Would be nice to not have to expose those, so a sidecar on the pod or service would be great.

I don't think any of my usecases different from the previous ones mentioned. The idea of not having to configure it by hand is very appealing.

**Comment by jakedgy** [comment]
We just built a prototype of an operator to implement a Tailscale load-balancer-controller, and we would like to open source it.

[`LoadBalancerClass`](https://github.com/kubernetes/enhancements/issues/1959) was promoted to beta in 1.22, so we can make it work side-by-side with existing cloud load-balancer-controllers.

We've implemented it as a `TailscaleProxy` custom resource that manages the full lifecycle of exposing some IP in a similar fashion to the proxy example.

It creates:
- Secrets to store node state information for stable Tailscale IPs for the life of the service
- A StatefulSet (no volumes, just using it for a stable pod identifier to store the node state in a corresponding secret)
- A `TailscaleEndpoint` per replica in the StatefulSet.  Whenever the Tailscale IPs become available (or a node disconnects), the `TailscaleEndpoint` is patched, which triggers an update of the Service `ExternalIPs` for all available TailscaleIPs for this service.

Since we are updating the `ExternalIPs`, you can use a separate external-dns controller to manage your DNS for the Tailscale IPs.

**Comment by evilhamsterman** [comment]
I particularly like the idea of a mutating webhook to automatically inject sidecars. It would be a simple way to exposing services securely to members of your team or for sharing. Especially for smaller or homelab deployments the overhead of running multiple tailscale daemons would be minimal. It wouldn't be a solution for all situations, but it would be a simple solution for many situations.

> 2. Providing inbound connectivity to a single pod e.g. for StatefulSet pods to communicate with each other
> 3. Providing outbound connectivity from pods
>    a. to Tailscale devices
>    b. other pods in the same cluster
>    c. cross-cluster connectivity
>    The obvious solution for 2&3 would be to automatically inject a Tailscale sidecar for each pod annotated with something like `ts.net/enable=true` with a help of a [Mutating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook). This has a few caveats though which makes this non-ideal.
>    
>    * It could potentially leak a lot of Tailscale nodes if not used carefully. This is something that [Ephemeral nodes](https://tailscale.com/kb/1111/ephemeral-nodes/) should help with.
>    * It'll end up potentially running many Tailscale nodes on the same machine, which would be expensive and wasteful. OTOH it would provide the ability to write pod-level [ACL](https://tailscale.com/kb/1018/acls/) policies with the help of [ACL tags](https://tailscale.com/kb/1068/acl-tags/).

**Comment by mkmik** [comment]
@jakedgy have you opensourced your controller? I'd like to try it out.

**Comment by clrxbl** [comment]
I decided to have a crack at this on my own at https://github.com/clrxbl/tailscale-svc-lb by creating a controller using the Python kopf framework. It's not great and could certainly use improvement (PRs welcome)., but as an initial prototype, this definitely does the job for me.

It takes inspiration from https://github.com/k3s-io/klipper-lb which is k3s' embedded service load balancer. For those that are unaware of it, it basically just deploys DaemonSets and runs iptables commands to forward requests to the cluster IP, similar to what I'm doing in this project.

[![asciicast](https://asciinema.org/a/smlS1PDekgvJBDuClsz9huMJy.svg)](https://asciinema.org/a/smlS1PDekgvJBDuClsz9huMJy)

**Comment by clrxbl** [comment]
Would like to mention that https://github.com/clrxbl/tailscale-svc-lb has gotten in a pretty good state after a major refactor from @bbetter173 so if anyone is still interested in an unofficial community Kubernetes operator for Tailscale, I would highly suggest checking it out now.

(I won't be posting more about it in this thread to avoid spam / self-advertisement)

**Comment by dvcrn** [comment]
Sorry for a bit OT talk, but I found this issue when searching for MagicDNS resolving within k8s. Did the instructions at https://github.com/tailscale/tailscale/issues/502#issuecomment-895342728 work for anyone recently? 

I'm trying to set this up with the fun domains (*.ts.net) and not having much luck to have coredns resolve tailscale names

**Comment by mayakacz** [comment]
**The Tailscale Kubernetes operator is now available in [alpha](https://tailscale.com/kb/1167/release-stages/#alpha) available for testing.** 🎉 

The Tailscale Kubernetes operator allows you to expose services in your Kubernetes cluster to your tailnet. You can expose a Kubernetes service over Tailscale in two ways: by making it a `LoadBalancer` type with the `tailscale` `loadBalancerClass`, or by annotating an existing service. Note that you’ll need the operator and proxy pods to use the latest unstable release of Tailscale, v.1.37.40 or later.

We’re still working on:
* Getting the operator working with Tailscale Funnel,
* Giving you a reverse proxy to route traffic back from the cluster to other nodes in your tailnet, and
* Creating an access proxy for apiserver.

**To get started, see the [documentation on the Kubernetes operator](https://tailscale.com/kb/1236/kubernetes-operator/) or check out the [demo video](https://www.youtube.com/watch?v=PGfJ18cQzho).**

Specific areas we’d like feedback on:
* Right now, there are no deployment methods other than applying the manifest file. How else would you like to deploy this?
* There are no automated updates. The operator and proxy pods will not update automatically to newer Tailscale releases as they become available. Would you want Tailscale to automatically update?
* What kinds of dashboards or metrics do you need for this operator, if any?

We look forward to hearing any feedback you have. Please add feedback to this GitHub issue.

Thank you for your ongoing support and feedback on Tailscale!

**Comment by clrxbl** [comment]
> **The Tailscale Kubernetes operator is now available in [alpha](https://tailscale.com/kb/1167/release-stages/#alpha) for testing.** 🎉
> 
> The Tailscale Kubernetes operator allows you to expose services in your Kubernetes cluster to your tailnet. You can expose a Kubernetes service over Tailscale in two ways: by making it a `LoadBalancer` type with the `tailscale` `loadBalancerClass`, or by annotating an existing service. Note that you’ll need the operator and proxy pods to use the latest unstable release of Tailscale, v.1.37.40 or later.
> 
> We’re still working on:
> 
> * Getting the operator working with Tailscale Funnel,
> * Giving you a reverse proxy to route traffic back from the cluster to other nodes in your tailnet, and
> * Creating an access proxy for apiserver.
> 
> **To get started, see the [documentation on the Kubernetes operator](https://tailscale.com/kb/1236/kubernetes-operator/) or check out the [demo video](https://www.youtube.com/watch?v=PGfJ18cQzho).**
> 
> Specific areas we’d like feedback on:
> 
> * Right now, there are no deployment methods other than applying the manifest file. How else would you like to deploy this?
> * There are no automated updates. The operator and proxy pods will not update automatically to newer Tailscale releases as they become available. Would you want Tailscale to automatically update?
> * What kinds of dashboards or metrics do you need for this operator, if any?
> 
> We look forward to hearing any feedback you have. Please add feedback to this GitHub issue.
> 
> Thank you for your ongoing support and feedback on Tailscale!

This is awesome news! Now to replace my hacked together operator with the official one.

> How else would you like to deploy this?
Tailscale should provide official Helm charts for this.

> Would you want Tailscale to automatically update?
I'd say that ideally Tailscale does not automatically update to prevent traffic disruption, and sets the appVersion in Tailscale's Helm chart to a specific Tailscale version. This should then be incremented everytime a Tailscale version gets released.

> What kinds of dashboards or metrics do you need for this operator, if any?
Generally this would mean introducing Prometheus metrics support to tailscaled, which would be a nice addition. Can't really think of much more other than basic traffic statistics like; connected clients, inbound/outbound packets per second & inbound/outbound bandwidth.

**Comment by jaredtbates** [comment]
This is great! Excited to try it out. We are starting to use Kustomize internally, so I'm pretty sure we can overlay the manifest file with our kustomizations as it is, but also, first party support for Kustomize would be welcome (https://github.com/kubernetes-sigs/kustomize/blob/master/examples/remoteBuild.md). As far as versioning, I agree with @clrxbl, ideally the version would be incremented in tandem with Tailscale, rather than auto-updating. And as far as metrics, OpenTelemetry support of some kind would be welcome - we use Datadog internally but can pipe OpenTelemetry into it easily (I think this is the same as Prometheus support?)

**Comment by almereyda** [comment]
It will also be interesting to see how the Headscale support of this turns
out. Currently it would be blocked by missing OAuth support there.

**Comment by evilhamsterman** [comment]
I second the Helm chart. That is the most popular way to deploy apps in Kubernetes. Kustomize is good to but it can be combined with Helm. As was said when you update Tailscale you also update the Helm chart, then when people want to upgrade they can pull the new chart and deploy it.

**Comment by dvcrn** [comment]
Great news! Thanks for adding this

I spent some time to rip out the ts-sidecars in my cluster and replaced them with this operator, so far worked great and no issues. 

Personally using kustomize and the current manifest was easy to use as is (taking out secrets and moving them into secretGenerator)

As others have mentioned, versioning is an issue. Currently I'd have to diff the manifests to see if anything has changed, so having some kind of key to make that easier to spot would be better. `tailscale/k8s-operator:unstable` isn't a great default, but fine during alpha

**Comment by RafalSkolasinski** [comment]
Really great news!

Question, is there plan for the operator to support 
a) subnet router -  to be able to access services in K8s cluster without adding them to tailnet
b) possibility to access from inside K8s cluster other machines on tailnet (ideally by magic dns)

**Comment by gnanderson** [comment]
This is so cool, thanks so much for this. To date we've been managing this ourselves with sidecars and/or mutating Kyverno policies, this will make life so much easier.

> Right now, there are no deployment methods other than applying the manifest file. How else would you like to deploy this?

Helm is a great start because it's so popular, however our preference would be OLM ( https://olm.operatorframework.io/docs/ ). We find issues with helm drift in our multi-tenancy situations so where we can we use OLM. It also opens the door to easy deployment into OpenShift clusters...

**Comment by uhthomas** [comment]
The operator is great, thank you so much. I have been dreaming of this for years.

Will the operator support the [Gateway API](https://gateway-api.sigs.k8s.io/)? A service load balancer endpoint is great and the additional flexibility with the Gateway API would be even better.

**Comment by gruberdev** [comment]
> Specific areas we’d like feedback on:
> * Right now, there are no deployment methods other than applying the manifest file. How else would you like to deploy this?

I'm deploying [using a Kustomize template](url)https://github.com/gruberdev/homelab/tree/main/apps/networking/tailscale/operator] onto my homelab cluster. 

I believe it is more intuitive than using Helm charts, would be optimal if Tailscale as an organization took responsability for maintaining a base deployment file using Kustomize. There are not enough resources or variables to customize in order to justify something complex as a Helm template.

> * There are no automated updates. The operator and proxy pods will not update automatically to newer Tailscale releases as they become available. Would you want Tailscale to automatically update?

No, automated updates to break on major versions will probably be a bad way to approach deployiong new features to the community. I believe informing and allowing people to make their own upgrades (at least through major versions)) is the more responsaible way to go.

> * What kinds of dashboards or metrics do you need for this operator, if any?

A Grafana dashboard that allows us to monitor how the flow of inbound connections is distributed across different services and namespaces wouild be intreresting. It would allow you to showcase tailscale capabilities, and at the same time, provide a good oversight of how the cluster is being used through Tailscale. 

I think providing a way to use Ingress with TLS termination is a good way to advance the project in order to improve the operator.

**Comment by bbetter173** [comment]
It would be a good idea to add the tailscale IP address to the Service's list of `externalIPs`. My use-case is exposing Traefik over Tailscale and having that IP address propagate to the Ingress resources that target it and currently I:

1. Expose the Traefik Service using annotations.
2. Note the IP address assigned to it in the Tailscale Console
3. Patch the `traefik` service with the new Tailscale IP

From here, the Kubernetes API takes care of everything else. If anyone plans to use this setup with externalDNS you'll probably want to add the `--target-net-filter=100.64.0.0/10` argument to ExternalDNS.

**Comment by ccesare** [comment]
If the `--accept-dns` and `--accept-routes` options could somehow be integrated to this operator (such that pods would use my tailnet DNS and have that traffic routed through my tailnet), my dreams would come true!

**Comment by uhthomas** [comment]
Has something changed with the operator recently? It crash loops for me now.

```
❯ k -n tailscale logs operator-958c74748-xxpgv
{"level":"info","ts":1679153663.8750303,"logger":"controller-runtime.metrics","msg":"Metrics server is starting to listen","addr":":8080"}
{"level":"info","ts":1679153663.8754432,"logger":"startup","msg":"Startup complete, operator running"}
{"level":"info","ts":1679153663.8760188,"msg":"Starting server","path":"/metrics","kind":"metrics","addr":"[::]:8080"}
{"level":"info","ts":1679153663.8761668,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.Service"}
{"level":"info","ts":1679153663.876277,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.StatefulSet"}
{"level":"info","ts":1679153663.8763,"msg":"Starting EventSource","controller":"service","controllerGroup":"","controllerKind":"Service","source":"kind source: *v1.Secret"}
{"level":"info","ts":1679153663.8763163,"msg":"Starting Controller","controller":"service","controllerGroup":"","controllerKind":"Service"}
2023/03/18 15:34:23 could not listen on :443: tsnet.Up: sending serve config: updating config: writing ServeConfig to StateStore: Secret "operator" is invalid: data[_serve/d692]: Invalid value: "_serve/d692": a valid config key must consist of alphanumeric characters, '-', '_' or '.' (e.g. 'key.name',  or 'KEY_NAME',  or 'key-name', regex used for validation is '[-._a-zA-Z0-9]+')
```

**Comment by devmattrick** [comment]
Thanks for the tip @uhthomas, that version worked for me.

If anyone is getting weird TLS errors like the following:

```
E0320 00:05:33.103280  228683 memcache.go:265] couldn't get current server API group list: Get "https://tailscale-operator.tailnet-name.ts.net/api?timeout=32s": remote error: tls: internal error
```

You'll need to enable HTTPS Certificates in your DNS settings for your Tailnet.

**Comment by dvcrn** [comment]
Noticed some smaller issues with my setup: Under certain circumstances, like when the cluster/nodes reboot, it spawns another operator. I have 3 now despite not touching my setup since my last comment. 

<img width="335" alt="image" src="https://user-images.githubusercontent.com/688326/227702565-96715044-36c2-4094-bacc-88b9ce8d841a.png">

Also today after an outage of my k8s provider, I noticed that all of my pods are disconnected from my tailnet (after operator-3 popped up). Key expiry is disabled so I had hoped it would just be able to reconnect without issues

**Comment by DentonGentry** [comment]
> I also have 6 tailscale-operator now in my tailnet

This means that two things are happening:
- the `tailscaled.state` is not being persisted. Depending on settings it could be stored either in /var/lib/tailscale/tailscaled.state or in the Kubernetes Secrets manager
- the authkey is not ephemeral, so it is creating nodes which don't delete themselves

The default for the operator is to store its state in the Kubernetes Secrets Manager, I'd recommend looking for errors about writing to secrets.

**Comment by uhthomas** [comment]
Not sure if this is in scope for the project, but it would be really cool if the operator could allow pods to use an exit node, or expose routes to the cluster. Right now I think a sidecar is required for those two situations.

**Comment by mayakacz** [comment]
> Creating an access proxy for apiserver.

A few weeks ago, we also added the ability to use the K8s operator as an auth proxy (sorry for not posting an update - though some of you in comments saw!).

You can use the K8s operator to access the Kubernetes control plane (kube-apiserver) over Tailscale. When a user tries to access the Kubernetes control plane, they will access it with the same user identity that they have in Tailscale.
To set this up, you need to deploy the Tailscale Kubernetes operator, update your kubeconfig locally, and ensure that the user has a cluster role that grants access to the Kubernetes control plane.

See the [documentation on the Kubernetes operator to getting this set up and running](https://tailscale.com/kb/1236/kubernetes-operator/#enabling-an-auth-proxy).

Specific areas we’d like feedback on:
* There is no way to define Tailscale ACLs that restrict Kubernetes RBAC, or map Tailscale ACLs to RBAC. How do you expect to handle user permissions in Kubernetes RBAC if accessing over Tailscale?
* Do you need the ability to record kubectl sessions?
* This operator only allows user@example.com to authenticate as themselves, not impersonate another user, like sre-service-account@example.com or just `sre`. Is impersonation something you need?

>> Giving you a reverse proxy to route traffic back from the cluster to other nodes in your tailnet

> Not sure if this is in scope for the project, but it would be really cool if the operator could allow pods to use an exit node, or expose routes to the cluster. Right now I think a sidecar is required for those two situations.

This is what we mean by reverse proxy - going from a Kubernetes service or pod out to a Tailscale. This isn't implemented yet; right now, these need a sidecar - [here is a sample sidecar to do this today](https://tailscale.com/kb/1185/kubernetes/#sample-sidecar).

**Comment by uhthomas** [comment]
Awesome, thank you. I cannot give enough praise for the official Tailscale x Kubernetes integration. I really had been dreaming of this for years and happy to see it finally become a reality.

Immediate feedback on:

> Is impersonation something you need?

Depending on the semantics, then yes absolutely. It would be great to have a separate service account and role to deploy to a cluster in CI for instance.

**Comment by shubb30** [comment]
Is it possible to configure environment variables in the tailscale proxy that is created for the service?  
We use an http proxy in our environment so we need to add `HTTP(S)_PROXY` env vars.

**Comment by shubb30** [comment]
Are there any plans to add HTTPS termination to the proxy setup?  Maybe adding a [Caddy](https://tailscale.com/kb/1190/caddy-certificates/) sidecar to the manifest?

**Comment by uhthomas** [comment]
> Are there any plans to add HTTPS termination to the proxy setup? Maybe adding a [Caddy](https://tailscale.com/kb/1190/caddy-certificates/) sidecar to the manifest?

No need for a sidecar. Tailscale supports HTTPS.

https://tailscale.com/kb/1153/enabling-https/

I don't believe this feature is enabled for the load balancer at the moment, though it would be nice! It may be worth opening an issue as a feature request to better track this.

**Comment by shubb30** [comment]
> Tailscale supports HTTPS.

That allows you to generate LE certificates for your hosts.  But in most cases, the pods running in Kubernetes are not running HTTPS, since that is usually handled by the Ingress or LoadBalancer.  Since Tailscale is bypassing both of those and going directly to the Kubernetes service, that means we would need to run an HTTPS web server on each pod in the workload.

**Comment by uhthomas** [comment]
> Is it possible to configure environment variables in the tailscale proxy that is created for the service? We use an http proxy in our environment so we need to add `HTTP(S)_PROXY` env vars.

I don't believe so, though I am confused. Does every container which needs network connectivity use the `HTTP(S)_PROXY` environment variable? If so, it sounds like this problem should be solved at the CNI level. Or are you asking about using a tailscale sidecar for tailnet connectivity? In which case, this may be helpful:

https://tailscale.com/kb/1185/kubernetes/

**Comment by shubb30** [comment]
In our network, our hosts need to go through an HTTP proxy to get to the Internet, which is needed for the TS proxy pods to register with Tailscale.  I managed to work around the issue by creating a custom image that is based on `tailscale:unstable` and adds the proxy environment variables, and now the TS proxy is able to register and it works.  

I think having the ability to add an annotation to the K8s service like `tailscale.com/proxy: http://proxyhost` which would tell the operator to add environment variables in the TS proxy pods.

**Comment by uhthomas** [comment]
> I've been working with the Tailscale Kubernetes operator and found that it currently creates child StatefulSet resources without any constraints on resource consumption. Additionally, there is no way to manually define the image tags used for the spawned resources.
> 
> It would be beneficial if the operator's configuration allowed users to define resource constraints (such as CPU and memory limits) as well as specify custom image tags for the child StatefulSets created by the operator. This would provide more control over the deployed resources and ensure better compatibility with specific environments or custom images.

I agree. It would be great to open an issue which tracks default resource requests and limits.

Could you clarify on `there is no way to manually define the image tags`? The `PROXY_IMAGE` environment variable should set the container image for the created stateful sets.

**Comment by dhess** [comment]
I've been banging my head against the operator for several hours now and can't figure out why it's not proxying a simple port 80/443 HTTP connection to an Argo CD service running inside the cluster. The operator itself seems to be working fine: it and the `argocd` machine both appear in my Tailnet, the tags are set up correctly, etc.

I can `tailscale ping` the `argocd` machine from an admin-authorized machine, and vice versa:

```
 /Applications/Tailscale.app/Contents/MacOS/Tailscale ping argocd
pong from argocd (100.67.18.68) via 10.8.11.64:52296 in 1ms
```

```
 kubectl exec --stdin=true --tty=true -n tailscale ts-argocd-server-8shb4-0 -- /usr/local/bin/tailscale --socket /tmp/tailscaled.sock ping sagan
Defaulted container "tailscale" out of: tailscale, sysctler (init)
pong from sagan (100.91.235.123) via 10.8.1.23:41641 in 1ms
```

`10.8.11.64` is the IP address of the k8s node on which the `argocd` proxy is running, and `10.8.1.23` is my laptop. Those IPs are directly connectable on my LAN.

However, if I try to ping the Tailscale IP of one from the other, going through the system network stack (i.e., `ping 100.67.18.68` from the laptop), no packets are returned; ditto for HTTP traffic from my laptop to the `argocd` Tailscale IP on port 80 and 443. The `argocd` proxy service's log shows the packets arriving and being accepted, but they never appear at the actual Argo CD pod:

```
2023/04/14 16:44:27 wgengine: idle peer [pvU4z] now active, reconfiguring WireGuard
2023/04/14 16:44:27 wgengine: Reconfig: configuring userspace WireGuard config (with 1/2 peers)
2023/04/14 16:44:27 magicsock: disco: node [pvU4z] d:984636d275129c4b now using 10.8.1.23:41641
2023/04/14 16:44:27 Accept: ICMPv4{100.91.235.123:0 > 100.67.18.68:0} 84 icmp ok
2023/04/14 16:44:28 Accept: ICMPv4{100.91.235.123:0 > 100.67.18.68:0} 84 icmp ok
2023/04/14 16:44:29 Accept: ICMPv4{100.91.235.123:0 > 100.67.18.68:0} 84 icmp ok
2023/04/14 16:44:38 Accept: TCP{100.91.235.123:49248 > 100.67.18.68:80} 64 tcp ok
2023/04/14 16:44:49 Accept: TCP{100.91.235.123:49249 > 100.67.18.68:443} 64 tcp ok
2023/04/14 16:45:01 Accept: TCP{100.91.235.123:49249 > 100.67.18.68:443} 64 tcp ok
2023/04/14 16:45:17 Accept: TCP{100.91.235.123:49249 > 100.67.18.68:443} 64 tcp ok
2023/04/14 16:45:49 Accept: TCP{100.91.235.123:49249 > 100.67.18.68:443} 48 tcp ok
```

I have another machine on my Tailnet that's not running on the LAN, but on a public IP address, and it's exhibiting the same issue as my laptop: it can ping `argocd` via Tailscale directly, but not through the system network stack, and HTTP packets are also being dropped somewhere despite being accepted by the `argocd` proxy.

The ACL I'm using is defined as follows:

```
{
			"action": "accept",
			"src": [
				"tag:argocd-client",
			],
			"dst": [
				"tag:argocd:80",
				"tag:argocd:443",
			],
		},
```

I allow port `80` and `443` as that's where the `argocd-server` Kubernetes service is listening, but allowing `*` makes no difference. Both the external machine and my laptop have the proper permissions to connect on these ports. I'm using the [annotation method](https://tailscale.com/kb/1236/kubernetes-operator/#exposing-a-service-using-annotations) on the `argocd-server` service so that I can use a custom tag and hostname, but I've also tried the [`LoadBalancer`](https://tailscale.com/kb/1236/kubernetes-operator/#exposing-a-service-using-loadbalancerclass) approach with the suggested `k8s` tag, and it's the same problem.

My Kubernetes cluster is running Talos v1.3.7, Kubernetes 1.26.3, and Cilium 1.13.1 with `kubeProxyReplacement=strict` (i.e., no `kube-proxy`). Cilium is running in tunneled (`vxlan`) mode with `ipam=kubernetes`. The cluster's `spec.podCIDR` is `10.244.0.0/16` (the Talos default) and the service CIDR is `10.96.0.0/12` (also the Talos default). Neither of these ranges overlaps with any LAN addresses, and neither is routable on the LAN. All of the Cilium connectivity tests pass, and all other pod-to-pod and node-to-node traffic is working fine. Originally I had Wireguard networking enabled in Cilium (for pod-to-pod traffic), and Kubespan enabled in Talos (for encrypted node-to-node traffic), but now all of that is disabled, yet the problem persists. (I've rebooted all the nodes after making the changes, as well, just to be safe.)

The Tailscale Kubernetes operator version and Tailscale container are `v1.39.147`.

One thing I've noticed that seems odd, is that in the Tailscale admin UI, the `argocd` machine only shows the `tailscaled` service running. I would have expected that it would also show the Argo CD service that it's proxying (HTTP on ports `80` and `443`). Should it?

Any suggestions would be greatly appreciated!

(Note: I had similar connectivity issues running a straight proxy &mdash; without the operator installed &mdash; by following the instructions [here](https://tailscale.com/kb/1185/kubernetes/#sample-proxy) and using the `v1.38.4` release of the Docker image, so I'm not sure this is, strictly speaking, an issue with the operator itself. I found out about the operator after spelunking a bit in the issue tracker.)

**Comment by uhthomas** [comment]
@dhess To avoid conflating this issue further, would you be able to open a new one, or maybe a discussion? I'd be happy to help.

**Comment by dhess** [comment]
@seh Maybe I'm being dense, but I don't understand your question. `100.67.18.68` is the Tailnet IP address assigned by Tailscale to the proxy service that's fronting the `argocd-server` service that's running in the Kubernetes cluster. In other words, it's the gateway address from my Tailnet to a service running inside the cluster and not otherwise available outside the cluster.

**Comment by seh** [comment]
> `100.67.18.68` is the Tailnet IP address assigned by Tailscale to the proxy service that's fronting the `argocd-server` service that's running in the Kubernetes cluster.

Oh, I couldn't glean that from what I had read, so I made a guess, given that the address was outside the other CIDR blocks you had mentioned.

**Comment by dhess** [comment]
> @dhess To avoid conflating this issue further, would you be able to open a new one, or maybe a discussion? I'd be happy to help.

Sure: #7868

**Comment by gruberdev** [comment]
> It would be a good idea to add the tailscale IP address to the Service's list of `externalIPs`. My use-case is exposing Traefik over Tailscale and having that IP address propagate to the Ingress resources that target it and currently I:
> 
> 1. Expose the Traefik Service using annotations.
> 2. Note the IP address assigned to it in the Tailscale Console
> 3. Patch the `traefik` service with the new Tailscale IP by adding it to the list of `externalIPs`
> 
> From here, the Kubernetes API takes care of everything else. If anyone plans to use this setup with externalDNS you'll probably want to add the `--target-net-filter=100.64.0.0/10` argument to ExternalDNS.

Inspired by this comment, [I did a few minor changes on the operator's code](https://github.com/tailscale/tailscale/commit/5282e26d5d22f290f12b793b4011e898820c8eca)

I've replaced the ExternalIP reference created by the operator to the Tailnet IP address, instead of default behavior of using your Tailnet's domain. You can build the container image yourself using this fork, or by using the container image: `docker.io/grubertech/tailscale-operator:latest`.

Now to make it work together with External-DNS and an Ingress controller, I noticed a few caveats:

- If you don't set `ExternalTrafficPolicy`, the requests are done using the Ingress controller's service IP Address, which are not a part of the Tailnet, thus making the requests fail.
- Your Ingress Controller has to support configuring the option `LoadBalancerclass` directly, only providing an option to configure it through annotations is not enough.

I also manage my resources using an open-source repository, so you can see how I configured them here:

- [External-DNS](https://github.com/gruberdev/homelab/blob/main/apps/argocd/base/core/external-dns.yaml)
- [Nginx Ingress Controller](https://github.com/gruberdev/homelab/blob/main/apps/argocd/base/core/nginx-ingress.yaml)

And here's a few examples as to how I'm configuring my resources:

- [Ingress](https://github.com/gruberdev/homelab/blob/main/apps/services/changedetection/base/ingress.yaml)
- [Service](https://github.com/gruberdev/homelab/blob/main/apps/services/changedetection/base/svc.yaml)

I hope this helps someone else, it's great to be able to centralize these configurations. I did not test if it works with CertManager yet, but I plan to post here as soon I learn this information.

A screenshot of the resulting services in a namespace of mine:
![](https://i.ibb.co/WfRjLr9/examplee.png)

**Comment by gruberdev** [comment]
> It would be a good idea to add the tailscale IP address to the Service's list of `externalIPs`. My use-case is exposing Traefik over Tailscale and having that IP address propagate to the Ingress resources that target it and currently I:
> 
> 1. Expose the Traefik Service using annotations.
> 2. Note the IP address assigned to it in the Tailscale Console
> 3. Patch the `traefik` service with the new Tailscale IP by adding it to the list of `externalIPs`
> 
> From here, the Kubernetes API takes care of everything else. If anyone plans to use this setup with externalDNS you'll probably want to add the `--target-net-filter=100.64.0.0/10` argument to ExternalDNS.

Inspired by this comment, [I did a few minor changes on the operator's code](https://github.com/tailscale/tailscale/commit/5282e26d5d22f290f12b793b4011e898820c8eca)

I've replaced the ExternalIP reference created by the operator to the Tailnet IP address, instead of default behavior of using your Tailnet's domain. You can build the container image yourself using this fork, or by using the container image:

- `docker.io/grubertech/tailscale-operator:latest`

Now to make it work together with External-DNS and an Ingress controller, I noticed a few caveats:

- You will need to change the controller's service `ExternalTrafficPolicy` to `Local`, this preserves the request header and avoids the request being redirected through kube-proxy, as it normally would when it is set to `Cluster` (default behavior in most controllers)
- The Ingress controller has to be exposed as one of the Tailnet's nodes, to ensure it can proxy the requests through the Tailscale network.
- This setup *might* work using CNAME and the Tailnet domain, without the need for the changes done to the operator, but I was unable to make it work. Only when providing an IP address to ExternalDNS, I was able to make it parse the requests succesfully. I believe it might be related to https://github.com/kubernetes-sigs/external-dns/issues/262.
- Your Ingress Controller has to support changing the parameter  `LoadBalancerClass` directly, only providing an option to configure it through annotations is not enough. You'll notice I'm using a fork of the Nginx Ingress, because the original one does not make that option available through its Helm chart.
- The TLS ingress termination works with Cloudflare and Nginx as the controller, the thing is that as long you own the domain, you can generate TLS certificates valid for only for your Tailnet applications.

I also manage my resources using an open-source repository, so you can see how I configured them here:

- [External-DNS](https://github.com/gruberdev/homelab/blob/main/apps/argocd/base/core/external-dns.yaml)
- [Nginx Ingress Controller](https://github.com/gruberdev/homelab/blob/main/apps/argocd/base/core/nginx-ingress.yaml)
- [CertManager](https://github.com/gruberdev/homelab/blob/main/apps/argocd/base/core/certmanager.yaml) <sub>[Cluster Issuer](https://github.com/gruberdev/homelab/blob/main/apps/networking/certmanager/clusterIssuer.yaml)

And here's a few examples as to how I'm configuring my resources:

- [Ingress](https://github.com/gruberdev/homelab/blob/main/apps/services/changedetection/base/ingress.yaml)
- [Service](https://github.com/gruberdev/homelab/blob/main/apps/services/changedetection/base/svc.yaml)
- [Certificate](https://github.com/gruberdev/homelab/blob/main/apps/services/changedetection/base/certificate.yaml)

I hope this helps someone else, it's great to be able to centralize these configurations.

A screenshot of the resulting services in a namespace of mine:
![](https://i.ibb.co/WfRjLr9/examplee.png)

**Comment by gnanderson** [comment]
@gruberdev 
> This setup might work using a CNAME and the Tailnet domain without making changes to the operator, but I couldn't make it work. Only when providing an IP address to ExternalDNS could I successfully parse the requests. I believe this might be related to https://github.com/kubernetes-sigs/external-dns/issues/262.

CNAME management using external-dns should work "out of the box" with a hostname annotation on the LoadBalancer service resource. You should just make sure your "txt prefix" is unique - so for example in the external-dns helm chart, set the `txtPrefix` value to something appropriate.

with that set, you should have no bother setting a CNAME in e.g. route53:

```
apiVersion: v1
kind: Service
metadata:
  annotations:
    tailscale.com/tags: 'tag:k8s-emissary'
    tailscale.com/hostname: 'friendly-hostname'
    external-dns.alpha.kubernetes.io/hostname: friendly-hostname.example.com,friendly-hostname.example2.com
    external-dns.alpha.kubernetes.io/ttl: "600"
```
external-dns values:

```
domainFilters:
  - example.com
  - example2.com

txtPrefix: do-devops
```

**Comment by cwiggs** [comment]
If I should open another issue for this please let me know.

I ran into an issue with the tailscale operator that I'm trying to figure out.  I have 2 clusters, each with the same version of tailscale-operator installed.  On the first cluster I wanted to get access to the api-server over tailscale so I annotated the `default.kubernetes` service with `tailscale.com/hostname` annotation, which seemed to work.  However after annotated the svc I noticed the documentation about enabling [auth proxy](https://tailscale.com/kb/1236/kubernetes-operator/#enabling-an-auth-proxy) and decided to try that route.  After enabling the AUTH_PROXY env var and trying to get it to work I decided to set it to false and use a different auth method. Everything works as expected on this cluster.

On this first cluster I'm still able to use the tailscale IP/DNS name with a separate auth method to get kubectl to talk to the cluster.  On the 2nd cluster I decided to try and use AUTH_PROXY instead of annotating the `default.kubernetes` SVC since the AUTH_PROXY method is in the docs.  After enabling AUTH_PROXY on the 2nd cluster I notice that port 443 is open, but I can't auth to the cluster with tailscale, or using oci cli (I'm able to auth to the first cluster with oci cli).

Here is the error I get when I try to connect to the 2nd cluster using tailscale as the server and oci cli as auth:

```
E0519 15:58:27.801815 3793607 memcache.go:265] couldn't get current server API group list: unknown
```

Questions:
* If I don't actually want to to use tailscale as auth but I still want to be able to reach the api-server over tialscale how should I go about doing that?  Annotate the `default.kubernetes` svc, auth_proxy, some other method?
* Any idea why the first cluster works but the 2nd cluster does not work.  I never annotated the `default.kubernetes` svc on the 2nd cluster so I assume that is why it's working on the first cluster; perhaps auth_proxy never worked for me?

Thanks.

**Comment by DentonGentry** [comment]
@cwiggs please open a new issue, this one is already long.

**Comment by cwiggs** [comment]
If we have questions about the tailscale operator where is the best place to ask?

I annotated a service and a pod with `ts-` spun up to create the tailscale connection.  I removed the annotation but the `ts-` pod sticks around even if it's killed, and the node is still active in my tailscale network.  How do I properly remove a "node" from the tailscale network if I added it via annotating a svc in k8s?

**Comment by uhthomas** [comment]
@cwiggs I would suggest opening a new issue.

**Comment by cwiggs** [comment]
> @cwiggs I would suggest opening a new issue.

Thanks, I opened this issue: https://github.com/tailscale/tailscale/issues/8192

**Comment by rodrigc** [comment]
@maisem @DentonGentry @danderson since the Tailscale k8s operator exists since https://github.com/tailscale/tailscale/pull/6713, I would suggest closing this issue.

I would also suggest that new issues can be created with the `OS-kubernetes` tag for issues found with the kubernetes operator. :)

**Comment by dvcrn** [comment]
Was anyone able to get the operator to play nicely with funnel? Last comment from february was 

> Getting the operator working with Tailscale Funnel,

but it's been a while since then. All I can think of at the top of my head would be to specify a custom tailscale container image that has funnel instructions setup

**Comment by maisem** [comment]
@dvcrn see https://github.com/tailscale/tailscale/issues/7895#issuecomment-1696227758

**Comment by dvcrn** [comment]
> @dvcrn see [#7895 (comment)](https://github.com/tailscale/tailscale/issues/7895#issuecomment-1696227758)

oh wonderful, thanks for that. I used https://github.com/valentinalexeev/tailscale-ingress-controller to spin up an ingress that works with funnel but great to see this coming baked in 👍

**Comment by gruberdev** [comment]
What @[shubb30](https://github.com/shubb30) said still holds to be true.

> > Tailscale supports HTTPS.
> 
> That allows you to generate LE certificates for your hosts. But in most cases, the pods running in Kubernetes are not running HTTPS, since that is usually handled by the Ingress or LoadBalancer. Since Tailscale is bypassing both of those and going directly to the Kubernetes service, that means we would need to run an HTTPS web server on each pod in the workload.

---

> > This setup might work using a CNAME and the Tailnet domain without making changes to the operator, but I couldn't make it work. Only when providing an IP address to ExternalDNS could I successfully parse the requests. I believe this might be related to [kubernetes-sigs/external-dns#262](https://github.com/kubernetes-sigs/external-dns/issues/262).
> 
> CNAME management using external-dns should work "out of the box" with a hostname annotation on the LoadBalancer service resource. You should just make sure your "txt prefix" is unique - so for example in the external-dns helm chart, set the `txtPrefix` value to something appropriate.
> 
> with that set, you should have no bother setting a CNAME in e.g. route53:
> 
> ```
> apiVersion: v1
> kind: Service
> metadata:
>   annotations:
>     tailscale.com/tags: 'tag:k8s-emissary'
>     tailscale.com/hostname: 'friendly-hostname'
>     external-dns.alpha.kubernetes.io/hostname: friendly-hostname.example.com,friendly-hostname.example2.com
>     external-dns.alpha.kubernetes.io/ttl: "600"
> ```
> 
> external-dns values:
> 
> ```
> domainFilters:
>   - example.com
>   - example2.com
> 
> txtPrefix: unique
> ```

Sadly, @gnanderson, it does not work. And the latest operator changes do not allow for the usage with `External-DNS` and an Ingress controller that is not Tailscale anymore, [which was how my setup was done earlier (and it worked with TLS termination on my custom hostname)])https://github.com/tailscale/tailscale/issues/502#issuecomment-1512871817).

I'm looking to avoid using funnel due to its bandwidth limits and would like to leverage my own hostname with `Cert-manager`. However, the operator's current handling of the `ExternalIP` field with multiple values complicates things, which isn't typically the norm. Any suggestions for addressing this would be appreciated.

**Comment by michaelbeaumont** [comment]
Just a note, I'm now successfully using `external-dns` with a custom domain and the operator with [the changes to put the tailscale IP in `status.loadBalancer.ingress`](https://github.com/tailscale/tailscale/pull/8817). You might want to use `--managed-record-types=A --managed-record-types=AAAA` to have `external-dns` only create A/AAAA records with the tailscale IP addresses and not CNAMEs with the `hostname` address. As an alternative, the `external-dns` option `--resolve-service-load-balancer-hostname` might work as well.

**Comment by gruberdev** [comment]
> Just a note, I'm now successfully using `external-dns` with a custom domain and the operator with [the changes to put the tailscale IP in `status.loadBalancer.ingress`](https://github.com/tailscale/tailscale/pull/8817). You might want to use `--managed-record-types=A --managed-record-types=AAAA` to have `external-dns` only create A/AAAA records with the tailscale IP addresses and not CNAMEs with the `hostname` address. As an alternative, the `external-dns` option `--resolve-service-load-balancer-hostname` might work as well.

Apologies, @michaelbeaumont. The setup did function as expected; the issue with HTTPS stemmed from a misconfiguration on my end, specifically within the Tailscale ACL policy and ingress controller settings.

That said, I truly value your suggestion. Interestingly, the setup also works with CNAME domains. While having multiple ExternalIPs is uncommon, it appears that ExternalDNS can handle it just fine with the configuration I've used.

**Comment by mayakacz** [comment]
### Additional functionality
We've added additional support for the Kubernetes operator:
* Support for ingress using an ingress resource. This allows you to expose services running in your cluster to resources on your tailnet. A single ingress can front multiple backends, and can also be used to expose a service over Tailscale Funnel.
* Support for using the API server proxy without auth headers. This allows you to use Tailscale to connect to the kube-apiserver without using Tailscale for auth.
* Support for egress using a proxy. This allows you to expose services on your tailnet to your cluster. Together with ingress, this allows for cross-cluster connectivity of services.

These updates are included in the feature documentation: https://tailscale.com/kb/1236/kubernetes-operator/.

The docs also include an updated list of limitations. Notably, we're working on certificate name matching issues (HTTPS support) for cluster egress and nftables support.

### Feature requests
We're going to close this bug. For any other feature requests, please upvote (using GitHub emojis) or open a new issue (see [tagged issues](https://github.com/tailscale/tailscale/labels/OS-kubernetes)). Some of the items we've already heard or think you might be interested in:
* A better deployment experience, e.g., a Helm chart or Kustomize. See https://github.com/tailscale/tailscale/issues/9222
* Metrics/monitoring, e.g., Grafana dashboard or Prometheus
* Automated updates of Tailscale, i.e. run tailscale update periodically / when the operator is redeployed
* Deploying subnet routers, exit nodes and app connectors for ingress to services and pods
* Allow customizing of the proxy pods for things like resource consumption, taints & tolerations etc.
* kubectl exec session recording / audit logging access to the kube apiserver
* Impersonation in API server proxy
* Map Tailscale identity & ACLs to Kubernetes RBAC
* Support for the Kubernetes Gateway API
* Visibility into whether a proxy is ready, e.g., custom Tailscale conditions to resources used to configure proxies

### Customer story
Lastly, if anyone is using the Kubernetes operator in an enterprise setting and this unblocked you to do something really cool, we'd love to include that in an announcement blog post or customer story. Please reach me at maya at tailscale.com if that applies to you.

Thanks!
